mat/gsvd.go

```diff
@@ -271,68 +275,76 @@ func (gsvd *GSVD) ZeroRTo(dst *Dense) *Dense {
 		dst.Slice(r, k+l, c+r-k-l, c).(*Dense).
 			Copy(b.Slice(r-k, l, c+r-k-l, c))
 	}
-	return dst
 }
 
 // SigmaATo extracts the matrix Σ₁ from the singular value decomposition, storing
 // the result in-place into dst. Σ₁ is size r×(k+l).
-// If dst is nil, a new matrix is allocated. The resulting SigmaA matrix is returned.
 //
-// SigmaATo will panic if the receiver does not contain a successful factorization.
-func (gsvd *GSVD) SigmaATo(dst *Dense) *Dense {
+// If dst is empty, SigmaATo will resize dst to be r×(k+l). If dst is
+// non-empty, SigmATo will panic if dst is not r×(k+l). SigmaATo will also
+// panic if the receiver does not contain a successful factorization.
+func (gsvd *GSVD) SigmaATo(dst *Dense) {
 	if !gsvd.succFact() {
 		panic(badFact)
 	}
 	r := gsvd.r
 	k := gsvd.k
 	l := gsvd.l
-	if dst == nil {
-		dst = NewDense(r, k+l, nil)
+	if dst.IsEmpty() {
+		dst.ReuseAs(r, k+l)
 	} else {
-		dst.reuseAsZeroed(r, k+l)
+		r2, c := dst.Dims()
+		if r2 != r || c != k+l {
+			panic(ErrShape)
+		}
+		dst.Zero()
 	}
 	for i := 0; i < k; i++ {
 		dst.set(i, i, 1)
 	}
 	for i := k; i < min(r, k+l); i++ {
 		dst.set(i, i, gsvd.s1[i])
 	}
-	return dst
 }
 
 // SigmaBTo extracts the matrix Σ₂ from the singular value decomposition, storing
 // the result in-place into dst. Σ₂ is size p×(k+l).
-// If dst is nil, a new matrix is allocated. The resulting SigmaB matrix is returned.
 //
-// SigmaBTo will panic if the receiver does not contain a successful factorization.
-func (gsvd *GSVD) SigmaBTo(dst *Dense) *Dense {
+// If dst is empty, SigmaBTo will resize dst to be p×(k+l). If dst is
+// non-empty, SigmBTo will panic if dst is not p×(k+l). SigmaBTo will also
+// panic if the receiver does not contain a successful factorization.
+func (gsvd *GSVD) SigmaBTo(dst *Dense) {
 	if !gsvd.succFact() {
 		panic(badFact)
 	}
 	r := gsvd.r
 	p := gsvd.p
 	k := gsvd.k
 	l := gsvd.l
-	if dst == nil {
-		dst = NewDense(p, k+l, nil)
+	if dst.IsEmpty() {
+		dst.ReuseAs(p, k+l)
 	} else {
-		dst.reuseAsZeroed(p, k+l)
+		r, c := dst.Dims()
+		if r != p || c != k+l {
+			panic(ErrShape)
+		}
+		dst.Zero()
 	}
 	for i := 0; i < min(l, r-k); i++ {
 		dst.set(i, i+k, gsvd.s2[k+i])
 	}
 	for i := r - k; i < l; i++ {
 		dst.set(i, i+k, 1)
 	}
-	return dst
 }
 
 // UTo extracts the matrix U from the singular value decomposition, storing
 // the result in-place into dst. U is size r×r.
-// If dst is nil, a new matrix is allocated. The resulting U matrix is returned.
 //
-// UTo will panic if the receiver does not contain a successful factorization.
-func (gsvd *GSVD) UTo(dst *Dense) *Dense {
+// If dst is empty, UTo will resize dst to be r×r. If dst is
+// non-empty, UTo will panic if dst is not r×r. UTo will also
```

Comment:
"When dst is non-empty, UTo will panic if dst is not r×r."

Reply:
Done.